### PR TITLE
feat(datasource/hex): extract deprecated versions

### LINKED ISSUE #26391: Ignore depreceated versions for `mix` dependencies
### Describe the proposed change(s).

_**Related Discussion: https://github.com/renovatebot/renovate/discussions/26362**_

Renovate does not extract information related to depreceated/retired versions while fetching the releases for mix deps using the hex datasource. This causes it to suggest PRs for retired versions
_Repro: https://github.com/RahulGautamSingh/reno-repro-mix/pull/1_

The depreceated/retired versions info is present in the [`retirements`](https://hex.pm/api/packages/ueberauth) field of the response returned from `hex` api. We need to use the `retirements` object to populate the `isDepreceated` field.

Changes need to made in this file:
https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/hex/schema.ts

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Extract retired versions from the response and use it to populate the `isDepreceated` field in release info.

## Context
Closes: https://github.com/renovatebot/renovate/issues/26391

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs a test

File: lib/modules/datasource/hex/schema.ts [hunk 1]
```diff
@@ -19,6 +20,13 @@ export const HexRelease = z
         inserted_at: z.string().optional(),
       }),
     ).refine((releases) => releases.length > 0, 'No releases found'),
+    retirements: z.record(
```
Comment (rarkins) [inline lib/modules/datasource/hex/schema.ts:23]: Is this field optional?

Comment (RahulGautamSingh) [reply lib/modules/datasource/hex/schema.ts:23]: Yes

Comment (rarkins) [reply lib/modules/datasource/hex/schema.ts:23]: Is it defined as optional in this zod schema?

Comment (rarkins) [reply lib/modules/datasource/hex/schema.ts:23]: I think make it optional, even if it's meant to be there

File: lib/modules/datasource/hex/schema.ts [hunk 2]
```diff
@@ -29,6 +37,10 @@ export const HexRelease = z
           release.releaseTimestamp = releaseTimestamp;
         }
 
+        if (is.plainObject(hexResponse.retirements[version])) {
```
Comment (rarkins) [inline lib/modules/datasource/hex/schema.ts:40]:
```suggestion
        if (is.plainObject(hexResponse.retirements?.[version])) {
```

**Comment by rarkins** [comment]
Where is this schema from?

**Comment by rarkins** [comment]
I can see this example: https://hex.pm/api/packages/ueberauth

But I can't find anywhere on the net where `retirements` is standardized

**Comment by RahulGautamSingh** [comment]
The docs are outdated hence we have to confirm the details by visting the api source code at https://github.com/hexpm/hexpm/blob/main/lib/hexpm_web/views/api

For our use-case we need to check the package-view and confirm the structure of the response:
https://github.com/hexpm/hexpm/blob/d6bfdb1ecae7f05b826b252be2e53f6a59da7a38/lib/hexpm_web/views/api/package_view.ex#L61

PS: [Here](https://github.com/hexpm/specifications/issues/31#issuecomment-950339364) is a comment from the owner confirming that the docs are outdated.

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by Cikey** [comment]
This breaks for hackney:

```
 WARN: Error fetching https://hex.pm/api/packages/hackney (repository=foobaR)
       "datasource": "hex",
       "packageName": "hackney",
       "err": {
         "message": "Schema error",
         "stack": "ZodError: Schema error\n    at Object.get error [as error] (/opt/containerbase/tools/renovate/37.137.3/node_modules/zod/lib/types.js:43:31)\n    at ZodEffects.parseAsync (/opt/containerbase/tools/renovate/37.137.3/node_modules/zod/lib/types.js:166:22)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Http.requestJson (/opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/util/http/index.ts:316:18)\n    at HexDatasource.getReleases (/opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/modules/datasource/hex/index.ts:45:34)\n    at /opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/util/cache/package/decorator.ts:119:18\n    at getRegistryReleases (/opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/modules/datasource/index.ts:77:15)\n    at fetchReleases (/opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/modules/datasource/index.ts:289:15)\n    at lookupUpdates (/opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:115:56)\n    at withLookupStats (/opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/workers/repository/process/fetch.ts:27:18)\n    at fetchDepUpdates (/opt/containerbase/tools/renovate/37.137.3/node_modules/renovate/lib/workers/repository/process/fetch.ts:72:30)\n    at /opt/containerbase/tools/renovate/37.137.3/node_modules/p-map/index.js:57:22",
         "issues": {
           "retirements": {
             "1.6.4": {"message": "Expected string, received null"},
             "1.6.6": {"message": "Expected string, received null"}
           }
         }
       }
bunyan: ERROR: Exception thrown from the "err" Bunyan serializer. This should never happen. This is a bug in that serializer function.
```

For whatever reasons, hackney has a `null` message: https://hex.pm/api/packages/hackney

**Comment by viceice** [comment]
@Cikey next time please open a new discussion instead of commenting on merged / closed PR's

- #26706

**Comment by Cikey** [comment]
I tried, but i didn't realize i had to use the `discussion` option to open a bug report 🤷🏻 next time..

**Pull request merged**